pkg/controlsvc/controlsvc_test.go

```diff
@@ -0,0 +1,647 @@
+					printExpectedError(t, err)
+				}
+			} else {
+				if err != nil {
+					printExpectedError(t, err)
+				}
+			}
+		})
+	}
+}
+
+func TestSockControlWriteToConn(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCon := mock_controlsvc.NewMockConn(ctrl)
+	mockUtil := mock_controlsvc.NewMockUtiler(ctrl)
+	mockCopier := mock_controlsvc.NewMockCopier(ctrl)
+
+	sockControl := controlsvc.NewSockControl(mockCon, mockUtil, mockCopier)
+
+	bridgeConnTestCases := []struct {
+		name          string
+		message       string
+		input         chan []byte
+		expectedCalls func()
+		expectedError bool
+		errorMessage  string
+	}{
+		{
+			name:    "without message and with error",
+			message: "",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New("write to conn chan error"))
+			},
+			expectedError: true,
+			errorMessage:  "write to conn chan error",
+		},
+		{
+			name:    "with message and with error",
+			message: "message",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, errors.New(writeToConnError))
+			},
+			expectedError: true,
+			errorMessage:  writeToConnError,
+		},
+		{
+			name:    "without message and error",
+			message: "",
+			expectedCalls: func() {
+				mockCon.EXPECT().Write(gomock.Any()).Return(0, nil)
+			},
+			expectedError: false,
+			errorMessage:  writeToConnError,
+		},
+	}
+
+	for _, testCase := range bridgeConnTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			testCase.expectedCalls()
+			c := make(chan []byte)
+			go func(c chan []byte) {
+				c <- []byte{7}
+				defer close(c)
+			}(c)
+
+			err := sockControl.WriteToConn(testCase.message, c)
+
+			if testCase.expectedError {
+				if err == nil && err.Error() != testCase.errorMessage {
+					printExpectedError(t, err)
+				}
+			} else {
+				if err != nil {
+					printExpectedError(t, err)
+				}
+			}
+		})
+	}
+}
+
+func TestSockControlClose(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCon := mock_controlsvc.NewMockConn(ctrl)
+	mockUtil := mock_controlsvc.NewMockUtiler(ctrl)
+	mockCopier := mock_controlsvc.NewMockCopier(ctrl)
+
+	sockControl := controlsvc.NewSockControl(mockCon, mockUtil, mockCopier)
+
+	errorMessage := "cannot close connection"
+
+	mockCon.EXPECT().Close().Return(errors.New(errorMessage))
+
+	err := sockControl.Close()
+	if err == nil && err.Error() != errorMessage {
+		printExpectedError(t, err)
+	}
+}
+
+func TestAddControlFunc(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCtrlCmd := mock_controlsvc.NewMockControlCommandType(ctrl)
+	mockNetceptor := mock_controlsvc.NewMockNetceptorForControlsvc(ctrl)
+	controlFuncTestsCases := []struct {
+		name          string
+		input         string
+		expectedError bool
+		errorMessage  string
+		testCase      func(msg string, err error)
+	}{
+		{
+			name:          "ping command",
+			input:         "ping",
+			expectedError: true,
+			errorMessage:  "control function named ping already exists",
```

Comment:
interesting, why does `ping` already exist? is this something left over from a previous test?